{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "import os\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import statistics\n",
    "import csv\n",
    "import matplotlib.pyplot as plt\n",
    "from pylab import figure, axes, title, show\n",
    "from scipy import stats\n",
    "import gzip\n",
    "import matplotlib.ticker as tkr\n",
    "from matplotlib import rc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# From the alignment, identify positions which are conserved in mammals and classify based on conservation in snakes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snakes = [\"Burmese_python\", \"Tiger_rattlesnake\", \"Brown-spotted_pit_viper\", \"Western_garter_snake\", \"Common_garter_snake\", \"Corn_snake\", \"Indian_cobra\", \"Eastern_brown_snake\", \"Tiger_snake\"]\n",
    "alignments = pd.read_csv(\"Z2_snakes_and_Cav_mammals_wDomains.tsv\", index_col=0, sep=\"\\t\", header=None)\n",
    "PWWP_Mam_Same = 0\n",
    "PWWP_Snake_Diff = 0\n",
    "ZFCW_Mam_Same = 0\n",
    "ZFCW_Snake_Diff = 0\n",
    "All_Same = []\n",
    "Most_Same = []\n",
    "Most_Diff =[]\n",
    "All_Diff = []\n",
    "CS_Diff = []\n",
    "PWWPrange = []\n",
    "ZFCWrange = []\n",
    "humanpos = 0\n",
    "mamcount = 0\n",
    "for column in alignments.columns:\n",
    "    snakeaa = {}\n",
    "    mammalaa = {}\n",
    "    PWWP = False\n",
    "    ZFCW= False\n",
    "    for index in alignments.index: \n",
    "        if index == \"Homo_sapiens\": \n",
    "            if alignments.loc[index, column] != \"-\": humanpos+= 1\n",
    "        elif index == \"Corn_snake\": \n",
    "            CSval = alignments.loc[index, column]\n",
    "        if index == \"ZF-CW\":\n",
    "            if alignments.loc[index, column]==\"TRUE\": \n",
    "                ZFCW = True\n",
    "                ZFCWrange.append(humanpos)\n",
    "        elif index == \"PWWP\":\n",
    "            if alignments.loc[index, column]==\"TRUE\": \n",
    "                PWWP = True\n",
    "                PWWPrange.append(humanpos)\n",
    "        elif index in snakes: \n",
    "            if alignments.loc[index, column] in snakeaa: snakeaa[alignments.loc[index, column]]+= 1\n",
    "            else: snakeaa[alignments.loc[index, column]] = 1\n",
    "        else: \n",
    "            if alignments.loc[index, column] in mammalaa: mammalaa[alignments.loc[index, column]]+= 1\n",
    "            else: mammalaa[alignments.loc[index, column]] = 1\n",
    "    total = 0\n",
    "    maxcount = 0\n",
    "    currmax = \"\"\n",
    "    for aa in mammalaa: \n",
    "        total += mammalaa[aa]\n",
    "        if mammalaa[aa] > maxcount: \n",
    "            maxcount = mammalaa[aa]\n",
    "            currmax = aa\n",
    "    if maxcount > (.95*total): \n",
    "        mammax = currmax\n",
    "        if mammax == \"-\": continue\n",
    "        else: \n",
    "            if ZFCW==True:\n",
    "                ZFCW_Mam_Same += 1\n",
    "            elif PWWP == True: \n",
    "                PWWP_Mam_Same += 1\n",
    "    else: continue\n",
    "    diff = 0\n",
    "    same = 0\n",
    "    for aa in snakeaa: \n",
    "        if aa != mammax: diff += snakeaa[aa]\n",
    "        else: same += snakeaa[aa]\n",
    "    diff = 0\n",
    "    same = 0\n",
    "    for aa in snakeaa: \n",
    "        if aa == \"-\": same += snakeaa[aa]\n",
    "        elif aa != mammax: diff += snakeaa[aa]\n",
    "        else: same += snakeaa[aa]\n",
    "    # If it's a CS-specific mutation, make that purple\n",
    "    if diff == 0: \n",
    "        All_Same.append(humanpos)\n",
    "    elif diff == 1 and CSval != mammax: \n",
    "        CS_Diff.append(humanpos)\n",
    "    elif diff < 5: \n",
    "        Most_Same.append(humanpos)\n",
    "    elif diff > 4 and diff < 9: \n",
    "        Most_Diff.append(humanpos)\n",
    "    elif diff == 9: \n",
    "        All_Diff.append(humanpos)\n",
    "        if ZFCW==True: ZFCW_Snake_Diff += 1\n",
    "        if PWWP==True: PWWP_Snake_Diff += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.patches import Rectangle\n",
    "#Make subplots\n",
    "plt.rcParams.update({'font.size': 6})\n",
    "fig = plt.figure(constrained_layout=True, dpi=1200)\n",
    "gs = fig.add_gridspec(15, 20)\n",
    "a1 = fig.add_subplot(gs[0:7, 0:6])\n",
    "a2 = fig.add_subplot(gs[0:7, 6:20])\n",
    "a3 = fig.add_subplot(gs[7:11, 0:5])\n",
    "a4 = fig.add_subplot(gs[11:15, 0:5])\n",
    "a5 = fig.add_subplot(gs[7:15, 5:20])\n",
    "#Plot binding assay data (5A) as a bar plot\n",
    "bindingassay = pd.read_csv(\"BindingAssay_Data.txt\",sep=\"\\t\")\n",
    "a1.bar(x=range(0,5), height=bindingassay.iloc[0:5, 4],color=\"C0\", zorder=1, yerr=[x*2 for x in bindingassay.iloc[0:5, 5]], capsize=2, label=\"Mouse ZCWPW2\\nn=2\", error_kw={\"linewidth\":.5})\n",
    "a1.bar(x=range(5,10), height=bindingassay.iloc[5:10, 4],color=\"C4\", zorder=2, yerr=[x*2 for x in bindingassay.iloc[5:10, 5]], capsize=2, label=\"Corn snake ZCWPW2\\nn=3\", error_kw={\"linewidth\":.5})\n",
    "a1.set_xticks(range(0, 10))\n",
    "a1.set_xticklabels(bindingassay.iloc[:, 0], rotation=70, fontsize=6)\n",
    "a1.set_ylabel(\"Band intensity \\nnormalized to K4/K36\", fontsize=6)\n",
    "a1.spines['right'].set_visible(False)\n",
    "a1.spines['top'].set_visible(False)\n",
    "\n",
    "#Add data points on 5A\n",
    "a1.scatter([x for x in range(0,10)], bindingassay.iloc[:, 1], color=\"black\", alpha=.7, s=7, zorder=3)\n",
    "a1.scatter([x-.1 for x in range(0,10)], bindingassay.iloc[:, 2], color=\"black\", alpha=.7, s=7, zorder=4)\n",
    "a1.scatter([x+.1 for x in range(0,10)], bindingassay.iloc[:, 3], color=\"black\", alpha=.7, s=7, zorder=5)\n",
    "a1.legend(loc=(.05, .75),fontsize=5)\n",
    "\n",
    "#Plot substitution data, calculated from above cell (5B)\n",
    "a2.spines['right'].set_visible(False)\n",
    "a2.spines['top'].set_visible(False)\n",
    "a2.spines['bottom'].set_visible(False)\n",
    "a2.spines['left'].set_visible(False)\n",
    "a2.vlines(x = All_Same, ymin = 2, ymax = 3,colors = 'darkgrey', label=\"Conserved in all snakes\", linewidth=.75)\n",
    "a2.vlines(x = Most_Same, ymin = 2, ymax = 3,colors = \"#ffb6db\", label = 'Substitution in <50% of snakes', linewidth=.75)\n",
    "a2.vlines(x = Most_Diff, ymin = 2, ymax = 3,colors = \"#ff6db6\", label = 'Substitution in >50% of snakes', linewidth=.75)\n",
    "a2.vlines(x = All_Diff, ymin = 2, ymax = 3,colors =  \"#490092\", label = 'Substitution in all snakes', linewidth=.75)\n",
    "\n",
    "#Add rectangles across PWWP and zf-CW range\n",
    "a2.add_patch(Rectangle((min(PWWPrange), 1.75), max(PWWPrange) - min(PWWPrange), 1.5, alpha=.2, color=\"grey\"))\n",
    "a2.add_patch(Rectangle((min(ZFCWrange), 1.75), max(ZFCWrange) - min(ZFCWrange), 1.5, alpha=.2, color=\"grey\"))\n",
    "a2.text(s=\"PWWP\",x=127,y=3.4,fontsize=6)\n",
    "a2.text(s=\"zf-CW\",x=43,y=3.4,fontsize=6)\n",
    "a2.hlines(2.5, 1, 356, color=\"black\", linewidth=1)\n",
    "a2.set_ylim(1, 6)\n",
    "a2.set_xlim(1, 360)\n",
    "a2.set_xlabel(\"ZCWPW2 amino acid position in human ortholog\", fontsize=6)\n",
    "a2.legend(title=\"Highly conserved in mammals\", ncol=2, fontsize=5, loc=\"upper center\")\n",
    "#plt.text(x=-44, y = 2.4, s=\"ZCWPW2\",fontsize=13)\n",
    "a2.set_yticks([], [])\n",
    "\n",
    "#Plot RMSD distributions for ZCWPW2 domains (5C-D)\n",
    "ZFCW = pd.read_csv(\"Z2_ZFCW_Hum_Snake.csv\")\n",
    "PWWP = pd.read_csv(\"Z2_PWWP_Hum_Snake.csv\")\n",
    "a3.spines['right'].set_visible(False)\n",
    "a3.spines['top'].set_visible(False)\n",
    "a3.spines['left'].set_visible(False)\n",
    "a4.spines['right'].set_visible(False)\n",
    "a4.spines['top'].set_visible(False)\n",
    "a4.spines['left'].set_visible(False)\n",
    "bin_s = np.arange(0, 1.5, .01)\n",
    "a4.hist(ZFCW.iloc[:, 3], bins=bin_s, color=\"#920000\", label=\"zf-CW domain\", alpha=.7)\n",
    "a4.hist(PWWP.iloc[:, 3], bins=bin_s, color=\"#009292\", label=\"PWWP domain\", alpha=.7)\n",
    "a4.set_xlabel(\"RMSD ($\\mathrm{\\AA}$) between ZCWPW2\\nstructure predictions\", fontsize=6)\n",
    "a4.tick_params(axis='both', which='major', labelsize=6)\n",
    "ZFCW = pd.read_csv(\"Z2_ZFCW_Hum_Mouse.csv\")\n",
    "PWWP = pd.read_csv(\"Z2_PWWP_Hum_Mouse.csv\")\n",
    "bin_s = np.arange(0, 1.5, .01)\n",
    "a3.set_title(\"Human-mouse comparison\",fontsize=6)\n",
    "a3.hist(ZFCW.iloc[:, 3], bins=bin_s, color=\"#920000\", label=\"zf-CW domain\", alpha=.8)\n",
    "a3.hist(PWWP.iloc[:, 3], bins=bin_s, color=\"#009292\", label=\"PWWP domain\", alpha=.8)\n",
    "a3.legend(fontsize=5)\n",
    "a3.set_yticks([],[])\n",
    "a4.set_yticks([],[])\n",
    "a3.set_xticks([],[])\n",
    "a4.set_title(\"Human-corn snake comparison\", fontsize=6)\n",
    "a5.spines['right'].set_visible(False)\n",
    "a5.spines['top'].set_visible(False)\n",
    "a5.spines['left'].set_visible(False)\n",
    "a5.spines['bottom'].set_visible(False)\n",
    "a5.set_yticks([],[])\n",
    "a5.set_xticks([],[])\n",
    "plt.savefig(\"Figure5.svg\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
